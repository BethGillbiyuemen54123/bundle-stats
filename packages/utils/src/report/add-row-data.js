import { getDelta, formatDelta } from '../utils/delta';
import { formatPercentage } from '../utils/format';
import { getMetricChanged } from '../metrics/get-metric-changed';
import { getMetricType } from '../metrics/get-metric-type';

export const addRowData = (entries, metricType, metricPrefix = '') => entries.map((entry) => {
  const { runs } = entry;

  const { biggerIsBetter, label, formatter } = getMetricType(
    metricPrefix ? [metricPrefix, entry.key].join('.') : entry.key,
    metricType,
  );

  return {
    ...entry,

    // Metric props
    biggerIsBetter,
    label,

    // Row props
    changed: getMetricChanged(runs),

    // Runs
    runs: runs.map((run, index) => {
      if (!run || typeof run.value === 'undefined' || run.value === null) {
        return null;
      }

      const { value, ...restRun } = run;
      const diff = (index < runs.length - 1) ? getDelta(runs[index + 1], run) : null;

      return {
        ...restRun,

        // run data
        value,
        displayValue: formatter(run.value),

        // diff data
        ...(diff !== null) ? {
          ...diff,
          displayDelta: formatDelta(diff.delta, formatter),
          displayDeltaPercentage: formatDelta(diff.deltaPercentage, formatPercentage),
        } : {},
      };
    }),
  };
});
