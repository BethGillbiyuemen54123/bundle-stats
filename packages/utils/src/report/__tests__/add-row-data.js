import { addRowData } from '../add-row-data';

describe('report / addRowData', () => {
  test('should add data', () => {
    const actual = addRowData([
      {
        key: 'metric1',
        type: 'METRIC_TYPE_DURATION',
        runs: [
          {
            value: 100,
          },
          {
            value: 110,
          },
        ],
      },
      {
        key: 'metric2',
        type: 'METRIC_TYPE_NUMBER',
        runs: [
          {
            value: 10,
          },
          null,
        ],
      },
      {
        key: 'metric4',
        type: 'METRIC_TYPE_SCORE',
        runs: [
          {
            value: 100,
          },
          {
            value: 100,
          },
        ],
      },
      {
        key: 'metric3',
        type: 'METRIC_TYPE_FILE_SIZE',
        runs: [
          null,
          {
            value: 10,
          },
        ],
      },
    ]);

    expect(actual).toMatchSnapshot();
  });
});
