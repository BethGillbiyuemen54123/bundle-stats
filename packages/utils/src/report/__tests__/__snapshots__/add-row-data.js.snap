// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report / addRowData should add data 1`] = `
Array [
  Object {
    "biggerIsBetter": false,
    "changed": true,
    "key": "metric1",
    "label": "metric1",
    "runs": Array [
      Object {
        "delta": -10,
        "deltaPercentage": -9.09090909,
        "displayDelta": "-10",
        "displayDeltaPercentage": "-9.09%",
        "displayValue": "100",
        "value": 100,
      },
      Object {
        "displayValue": "110",
        "value": 110,
      },
    ],
    "type": "METRIC_TYPE_DURATION",
  },
  Object {
    "biggerIsBetter": false,
    "changed": true,
    "key": "metric2",
    "label": "metric2",
    "runs": Array [
      Object {
        "delta": 10,
        "deltaPercentage": 100,
        "displayDelta": "+10",
        "displayDeltaPercentage": "+100%",
        "displayValue": "10",
        "value": 10,
      },
      null,
    ],
    "type": "METRIC_TYPE_NUMBER",
  },
  Object {
    "biggerIsBetter": false,
    "changed": false,
    "key": "metric4",
    "label": "metric4",
    "runs": Array [
      Object {
        "delta": 0,
        "deltaPercentage": 0,
        "displayDelta": "0",
        "displayDeltaPercentage": "0%",
        "displayValue": "100",
        "value": 100,
      },
      Object {
        "displayValue": "100",
        "value": 100,
      },
    ],
    "type": "METRIC_TYPE_SCORE",
  },
  Object {
    "biggerIsBetter": false,
    "changed": true,
    "key": "metric3",
    "label": "metric3",
    "runs": Array [
      null,
      Object {
        "displayValue": "10",
        "value": 10,
      },
    ],
    "type": "METRIC_TYPE_FILE_SIZE",
  },
]
`;
